tests/test_kb_auth.py

```diff
@@ -0,0 +1,123 @@
+async def test_create():
+    auth_client = await KBaseAuth.create(AUTH_URL)
+    assert isinstance(auth_client, KBaseAuth)
+
+
+async def test_create_non_auth_url():
+    with pytest.raises(IOError, match="does not appear to be the KBase Authentication Service"):
+        await KBaseAuth.create("https://github.com")
+
+
+async def test_create_incorrect_url():
+    with pytest.raises(IOError, match="Error from KBase auth server"):
+        await KBaseAuth.create(AUTH_URL + "wrong")
+
+
+async def test_non_json_error():
+    with pytest.raises(IOError, match="Non-JSON response from KBase auth server"):
+        await KBaseAuth.create("https://kbase.us/services")
+
+
+# TODO: test other service fail modes
+
+
+async def test_get_user(auth_client: KBaseAuth):
+    assert await auth_client.get_user(TEST_TOKEN) == TEST_USER
+
+
+async def test_get_user_from_cache(auth_client: KBaseAuth):
+    # patch this to act as a spy, should only be called once with the caching done.
+    with patch("staging_service.kb_auth_client._get", wraps=_get) as spy_get:
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        # make sure it's only called once
+        spy_get.assert_awaited_once()
+
+
+async def test_get_user_drop_cache():
+    auth_client = await KBaseAuth.create(AUTH_URL, cache_expiration=1)
+    with patch("staging_service.kb_auth_client._get", wraps=_get) as spy_get:
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        # make sure it's only called once at this point
+        spy_get.assert_awaited_once()
+        await asyncio.sleep(1)
+
+        # cache should have dropped the user
+        result = await auth_client.get_user(TEST_TOKEN)
+        assert result == TEST_USER
+
+        # expect this to be 2
+        assert spy_get.call_count == 2
+
+
+# TODO: test cache max size?
+
+
+@pytest.mark.parametrize("token", ["", None])
+async def test_get_user_no_token(auth_client: KBaseAuth, token: str | None):
+    with pytest.raises(ValueError, match="token is required"):
+        await auth_client.get_user(token)
+
+
+async def test_get_user_fail(auth_client: KBaseAuth):
+    with pytest.raises(InvalidTokenError, match="token is invalid"):
+        await auth_client.get_user("some_token")
+
+
+async def test_is_valid_user(auth_client: KBaseAuth):
+    assert await auth_client.is_valid_user(VALID_TEST_USER, TEST_TOKEN)
+
+
+async def test_is_valid_user_not_exist(auth_client: KBaseAuth):
+    assert not await auth_client.is_valid_user(NOT_REAL_USER, TEST_TOKEN)
+
+
+async def test_is_valid_user_invalid(auth_client: KBaseAuth):
+    with pytest.raises(InvalidUserError, match="Illegal character in user name"):
+        await auth_client.is_valid_user(INVALID_USER, TEST_TOKEN)
```

Comment:
codecov is also saying this doesn't exercise the if block in the error handler which seems impossible

Reply:
Looks the same to me: https://app.codecov.io/gh/kbase/staging_service/pull/218?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=kbase

Reply:
See above